web/src/stream_data.ts

```diff
@@ -8,10 +8,30 @@ import * as people from "./people";
 import * as settings_config from "./settings_config";
 import * as settings_data from "./settings_data";
 import * as sub_store from "./sub_store";
+import type {
+    ApiStreamSubscription,
+    ExtraStreamAttrs,
+    NeverSubscribedStream,
+    Stream,
+    StreamSubscription,
+} from "./sub_store";
 import * as user_groups from "./user_groups";
 import {user_settings} from "./user_settings";
 import * as util from "./util";
 
+type StreamInitParams = {
+    subscriptions: ApiStreamSubscription[];
+    unsubscribed: ApiStreamSubscription[];
+    never_subscribed: NeverSubscribedStream[];
+    realm_default_streams: Stream[];
+};
+
+// Type for the parameter of `create_sub_from_server_data` function.
+type StreamAttrs =
+    | (NeverSubscribedStream & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">)
+    | (ApiStreamSubscription & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">)
+    | (Stream & {stream_weekly_traffic: number | null; subscribers: number[]; subscribed: boolean}); // This type is used from `create_streams` function.
```

Comment:
Because `NeverSubscribedStream`, `ApiStreamSubscription`, `Stream` are all extended with `subscribed`, and that `previously_subscribed` is not optional for the first two types while defined optional on `ExtraStreamAttrs`, we can perhaps refactor the type like this:
```
    (
      | (NeverSubscribedStream & {previously_subscribed: boolean})
      | (ApiStreamSubscription & {previously_subscribed: boolean})
      | (Stream & {stream_weekly_traffic: number | null; subscribers: number[]})
    ) & {subscribed: boolean};
```

Reply:
Yeah that seems a bit cleaner.

Reply:
Would it make sense for `previously_subscribed` to be an always-present parameter? I guess it doesn't have a clear meaning for a stream you're currently subscribed to, so maybe that's a bad idea.

Reply:
That refactored type looks much better to me thanks @PIG208, as for `previously_subscribed` field I am not so sure about making it non-optional.

Reply:
Yeah this sounds good to me, did this refactoring in a separate commit 👍